Fix #167: Styled thick horizontal lines are scrambled

cmb69 opened the pull request:
Thick lines are drawn by gdImageFilledRectangle(), which iterates over
the x ordinate first (Z order) to apply the style pattern. While this works
fine for vertical and diagonal lines, it obviously fails for horizontal
lines, which have to be iterated over in N order.

To fix this bug, we introduce the helper _gdImageFilledRectangle() to
avoid having to add another parameter to gdImageFilledRectangle() or
to add another member to the gdImageStruct (both might break ABI
compatibility).

vapier commented on tests/gdimageline/github_bug_167.c:
```diff
@@ -0,0 +1,62 @@
+#include "gd.h"
```
add a comment header explaining what's going on and such

vapier commented on tests/gdimageline/github_bug_167.c:
```diff
@@ -0,0 +1,62 @@
+
+    while (i < 16 * thickness) {
+        style[i++] = color;
+    };
```
no spurious semi-colons ... comes up a lot below

vapier commented on tests/gdimageline/github_bug_167.c:
```diff
@@ -0,0 +1,62 @@
+{
+	gdImagePtr im;
+	int error = 0, black;
+    char path[1024];
```
you seem to have mixed indentation -- use tabs everywhere

vapier commented on tests/gdimageline/github_bug_167.c:
```diff
@@ -0,0 +1,62 @@
+    gdImageLine(im, 700, 100, 700, 600, gdStyled);
+    gdImageLine(im, 700, 600, 200, 100, gdStyled);
+
+    sprintf(path, "%s/gdimageline/github_bug_167_exp.png", GDTEST_TOP_DIR);
```
don't use sprintf -- look at how other tests use gdAssertImageEqualsToFile now

vapier commented:
you'll need to fix the Travis build failures first ;)

vapier commented:
also should update .gitignore in that subdir to ignore the new compiled test file

cmb69 commented:
I'll look into fixing what you have commented. Thanks.

> you'll need to fix the Travis build failures first ;)

Hm, these seem to be unrelated to the PR, see <https://travis-ci.org/libgd/libgd/jobs/136766458#L3773-L3775>, but I'm not sure.

vapier commented:
sorry, when i looked at the history, i thought i saw a green build before yours.  i've pushed a few commits, so both GD-2.2 & master should be green again.

note: when fixing things, you should squash all your commits into one.

cmb69 commented:
I have addressed all issues, including the squash (hopefully, I didn't make too much of mess; had to do a forced push). The tests are passing now also.

Please review. :)

Oh, and I did base my branch on master – probably, that should have been GD-2.2. Shall I fix that?

vapier commented:
imo, all work should be done first in master, and any fixes we think/want in an older branch should be cherry picked there

vapier commented on tests/gdimageline/github_bug_167.c:
```diff
@@ -0,0 +1,73 @@
+
+	To have the same pattern style, it has to be adapted according to the
+	thickness (i.e. the pixels have to be repeated thickness times), which
+	happens in setStyleAndThickness().
```
i'm lazy and often like to stick a link to the github issue at the end of the comment block too

vapier commented:
code looks fine, but i'd like it more if @pierrejoye could take a look

pierrejoye commented on src/gd.c:
```diff
@@ -2339,8 +2342,8 @@ BGD_DECLARE(void) gdImageRectangle (gdImagePtr im, int x1, int y1, int x2, int y
-BGD_DECLARE(void) gdImageFilledRectangle (gdImagePtr im, int x1, int y1, int x2, int y2,
-		int color)
+static void _gdImageFilledRectangle (gdImagePtr im, int x1, int y1, int x2, int y2,
+		int color, int horizontal)
```
I would rename this function to

_gdFilledPatternRectangle to avoid confusion.

The last parameter should be called direction or axis to reflect what it actually does.

Only to know what it is in 3y from now. 😎

cmb69 replied:
I'm not sure if _gdFilledPatternRectangle() would be a good choice, because this function actually draws *all* filled rectangles (the original gdImageFilledRectangle() being only a small wrapper on top of it). What about _gdImageFilledRectangleImp() or maybe _drawFilledRectangle()?

Renaming the last parameter to direction or axis is certainly more clarifying if we use an enum (or simple constants) instead of a bool. Not sure, if there's already something useful in libgd what can be reused – will have a look.

Another option would be to have two functions, say _gdImageFilledHRectangle() and _gdImageFilledVRectangle() instead. That would suit nicely to [gdImageHLine()](https://github.com/libgd/libgd/blob/gd-2.2.1/src/gd.c#L1247) and [gdImageVLine()](https://github.com/libgd/libgd/blob/gd-2.2.1/src/gd.c#L1266), for instance. The drawback would be yet more code duplication unless we'd use a macro (function inlining might not be available?)

pierrejoye replied:
Oh Right, I missed that the new functions are also used for generic rectangles.

You are right too about H and V as I can see needs for that in other area where styles or patterns are used.

cmb69 commented:
@vapier wrote:

> imo, all work should be done first in master, and any fixes we think/want in an older branch should be cherry picked there

Fine, thanks. That info should also be added to CONTRIBUTING, see #142.

pierrejoye commented:
I usually do the other way.

Fixes go to current stable branch anyway which is 2.2 and merge to master.
I find it easier and make safer not to forget a fix for the stable branch.

Any argument to do it the other way.

For new features a PR should be done and be discussed and indeed only
master can get new features (currently targeting 2.3)

Cheers
On Jun 11, 2016 4:23 PM, "Christoph M. Becker" <notifications@github.com>
wrote:

> @vapier <https://github.com/vapier> wrote:
>
> imo, all work should be done first in master, and any fixes we think/want
> in an older branch should be cherry picked there
>
> Fine, thanks. That info should also be added to CONTRIBUTING, see #142
> <https://github.com/libgd/libgd/issues/142>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/libgd/libgd/pull/222#issuecomment-225346899>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AARPKGZfPVBM1jrdPzU31BjmIjVeS8Yhks5qKn6NgaJpZM4IzLa0>
> .
>

pierrejoye commented:
Thanks for your work :)

I think you can go ahead after changing the names to_gdImageFilledHRectangle() and _gdImageFilledVRectangle() 

2.2 and master should have the fix.

If no big issue left we can most likely go with 2.2.2 this week.

cmb69 commented:
Fine, I'll try to fix the PR today.

pierrejoye commented:
Looks good to me 🤗

Go ahead with the merge (if you can do php 5.6+ as well, go ahead too:)

Thanks for the awesome work:)

cmb69 commented:
Fine, I'll do the merge ASAP, and also commit respective changes to PHP 5.6+ as well.

cmb69 commented:
I have committed c2b91db via git am, so I'm manually closing the PR.

pierrejoye commented:
Thanks :)
Btw only to be sure, do you have the permissions to close/merge PRs? Should
:)
On Jun 15, 2016 10:59 PM, "Christoph M. Becker" <notifications@github.com>
wrote:

> Closed #222 <https://github.com/libgd/libgd/pull/222>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/libgd/libgd/pull/222#event-693458649>, or mute the
> thread
> <https://github.com/notifications/unsubscribe/AARPKO0Etbe7n_tySm6pAtKDgdcGgLyaks5qMCFLgaJpZM4IzLa0>
> .
>

cmb69 commented:
I have permission to close PRs, and I assume that I also have merge permissions. :)
